src/base/bittorrent/sessionimpl.cpp

```diff
@@ -2711,8 +2711,39 @@ bool SessionImpl::addTorrent_impl(const TorrentDescriptor &source, const AddTorr
     if (m_loadingTorrents.contains(id) || (infoHash.isHybrid() && m_loadingTorrents.contains(altID)))
         return false;
 
-    if (findTorrent(infoHash))
+    if (Torrent *torrent = findTorrent(infoHash))
+    {
+        // a duplicate torrent is being added
+
+        if (hasMetadata)
+        {
+            // Trying to set metadata to existing torrent in case if it has none
+            torrent->setMetadata(*source.info());
+        }
+
+        if (!isMergeTrackersEnabled())
+        {
+            LogMsg(tr("Detected an attempt to add a duplicate torrent. Existing torrent: %1. Result: %2")
+                    .arg(torrent->name(), tr("Merging of trackers is disabled")));
+            return false;
+        }
+
+        const bool isPrivate = torrent->isPrivate() || (hasMetadata && source.info()->isPrivate());
```

Comment:
Isn't `isPrivate` true only when `hasMetadata` is also true https://github.com/qbittorrent/qBittorrent/pull/20951#discussion_r1640867755? (check for `hasMetadata` is superfluous)

Reply:
Checking for `hasMetadata` is required for safe access to `TorrentDescriptor::info()` (which is `optional`).